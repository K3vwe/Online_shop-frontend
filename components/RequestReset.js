import React, { useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import Error from "./ErrorMessage";
import Form from "./styles/Form";

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

const RequestReset = () => {
  const [email, setEmail] = useState("");
  const [reset, { loading, error, called }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: { email },
    }
  );

  const saveToState = (event) => {
    const { value } = event.target;
    setEmail(value);
  };
  return (
    <div>
      <Form
        method="POST"
        data-test="form"
        onSubmit={async (e) => {
          e.preventDefault();
          await reset().catch((error) => {
            console.log(error);
          });
          setEmail("");
        }}
      >
        <fieldset disabled={loading} aria-busy={loading}>
          <h2>Request a password reset</h2>
          <Error error={error} />
          {!error && !loading && called && (
            <p>Success! Check your email for a reset link!</p>
          )}
          <label htmlFor="email">
            Email
            <input
              type="email"
              name="email"
              plaeholder="enter your email here"
              value={email}
              onChange={saveToState}
            />
          </label>
          <button className="button"> Request Reset</button>
        </fieldset>
      </Form>
    </div>
  );
};

export default RequestReset;
