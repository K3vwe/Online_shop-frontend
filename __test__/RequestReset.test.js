import { mount } from "enzyme";
import wait from "waait";
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from "../components/RequestReset";
import { MockedProvider } from "@apollo/react-testing";
import toJson from "enzyme-to-json";

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: "testMail@mail.com" },
    },
    result: {
      data: { requestReset: { message: "success", __typename: "Message" } },
    },
  },
];

describe("<RequestReset/>", () => {
  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );

    const form = wrapper.find('form[data-test="form"]');
    expect(toJson(form)).toMatchSnapshot();
  });

  it("calls the mutation", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );

    // Simulate typing an email
    wrapper.find("input").simulate("change", {
      target: { name: "email", value: "testMail@mail.com" },
    });

    // Simulate a submit action
    wrapper.find("form").simulate("submit");
    await wait();
    wrapper.update();
    expect(wrapper.find("p").text()).toContain(
      "Success! Check your email for a reset link!"
    );
  });
});
