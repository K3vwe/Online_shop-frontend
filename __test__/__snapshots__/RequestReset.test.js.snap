// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 ekayYU"
  data-test="form"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Request a password reset
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        plaeholder="enter your email here"
        type="email"
        value=""
      />
    </label>
    <button
      className="button"
    >
       Request Reset
    </button>
  </fieldset>
</form>
`;
